File: Graphs/BFS.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution {
  public:
    // Function to return Breadth First Traversal of given graph.
    vector<int> bfsOfGraph(int V, vector<int> adj[]) {
        vector<int>bfs;
        queue<int>qt;
        qt.push(0);
        vector<int>vis(V,0);
        vis[0]=1;
        while(!qt.empty()){
            int curr = qt.front();
            qt.pop();
            bfs.push_back(curr);
            for(auto x: adj[curr]){
                if(!vis[x]){
                    qt.push(x);
                    vis[x]=1;
                }
            }
        }
        return bfs;
    }
};

void addEdge(vector <int> adj[], int u, int v) {
    adj[u].push_back(v);
    adj[v].push_back(u);
}

void printAns(vector <int> &ans) {
    for (int i = 0; i < ans.size(); i++) {
        cout << ans[i] << " ";
    }
}

int main() 
{
    vector <int> adj[6];
    
    addEdge(adj, 0, 1);
    addEdge(adj, 1, 2);
    addEdge(adj, 1, 3);
    addEdge(adj, 0, 4);

    Solution obj;
    vector <int> ans = obj.bfsOfGraph(5, adj);
    printAns(ans);

    return 0;
}

// Output: 0 1 4 2 3

// Time Complexity: O(N) + O(2E), Where N = Nodes, 2E is for total degrees as we traverse all adjacent nodes.

// Space Complexity: O(3N) ~ O(N), Space for queue data structure visited array and an adjacency list

File: Graphs/BFS.java
import java.util.*;

class bfsSolution {
    // Function to return Breadth First Traversal of given graph.
    public List<Integer> bfsOfGraph(int V, List<List<Integer>> adj) {
        List<Integer> bfs = new ArrayList<>();
        Queue<Integer> qt = new LinkedList<>();
        boolean[] visited = new boolean[V];
        
        qt.add(0);
        visited[0] = true;
        
        while (!qt.isEmpty()) {
            int curr = qt.poll();
            bfs.add(curr);
            for (int x : adj.get(curr)) {
                if (!visited[x]) {
                    qt.add(x);
                   visited[x] = true;
                }
            }
        }
        return bfs;
    }
}

public class BFS {
    public static void addEdge(List<List<Integer>> adj, int u, int v) {
        adj.get(u).add(v);
        adj.get(v).add(u);
    }

    public static void printAns(List<Integer> ans) {
        for (int i : ans) {
            System.out.print(i + " ");
        }
    }

    public static void main(String[] args) {
        int V = 6;
        List<List<Integer>> adj = new ArrayList<>(V);
        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }
        
        addEdge(adj, 0, 1);
        addEdge(adj, 1, 2);
        addEdge(adj, 1, 3);
        addEdge(adj, 0, 4);

        bfsSolution obj = new bfsSolution();
        List<Integer> ans = obj.bfsOfGraph(V, adj);
        printAns(ans);
    }
}